boa/src/syntax/parser/mod.rs

```diff
@@ -1305,44 +1380,62 @@ impl Parser {
     /// <https://tc39.github.io/ecma262/#prod-CallExpression>
     fn read_call_expression(&mut self, first_member_expr: Node) -> ParseResult {
         let mut lhs = first_member_expr;
-        if let Ok(true) =
-            self.next_if_skip_lineterminator(TokenKind::Punctuator(Punctuator::OpenParen))
+        if self
+            .next_if_skip_lineterminator(TokenKind::Punctuator(Punctuator::OpenParen))
+            .is_some()
         {
             let args = self.read_arguments()?;
             lhs = Node::Call(Box::new(lhs), args);
         } else {
-            panic!("CallExpression MUST start with MemberExpression.");
+            let next_token = self
+                .next_skip_lineterminator()
+                .ok_or(ParseError::AbruptEnd)?;
+            return Err(ParseError::Expected(
+                vec![TokenKind::Punctuator(Punctuator::OpenParen)],
+                next_token.clone(),
+                Some("call expression"),
+            ));
         }
 
-        while let Ok(tok) = self.peek_skip_lineterminator() {
+        while let Some(tok) = self.peek_skip_lineterminator().cloned() {
```

Comment:
Could we somehow avoid this? It seems that Rust gives a "multiple mutable reference" error when you return a read-only reference from a mutable reference more than once. Is this something the compiler should understand it's safe?

Reply:
> Could we somehow avoid this? It seems that Rust gives a "multiple mutable reference" error when you return a read-only reference from a mutable reference more than once. Is this something the compiler should understand it's safe?

I found a way. If you take a `&[Token]` instead of an owning `Vec<Token>` in `Cursor` and you annotate the returning references with `'a`, so `&'a Token` in functions that return it like `next`, `next_if`, etc. (also in Parser) you can eliminate the `clone`. 
BTW I'm still new to rust lifetimes so there are probably better ways of doing this.

Reply:
Yeah I think this is correct, its the same thing I would have done.